drivers/google/compute_util_test.go

```diff
@@ -46,27 +77,125 @@ func TestPortsUsed(t *testing.T) {
-		{"both ports opened", []*raw.FirewallAllowed{{IPProtocol: "tcp", Ports: []string{"2376", "3376"}}}, []string{"2376", "3376"}, map[string][]string{}},
-		{"more ports opened", []*raw.FirewallAllowed{{IPProtocol: "tcp", Ports: []string{"2376", "3376", "22", "1024-2048"}}}, []string{"2376", "3376"}, map[string][]string{}},
-		{"additional missing", []*raw.FirewallAllowed{{IPProtocol: "tcp", Ports: []string{"2376", "2377/tcp"}}}, []string{"2377/udp", "80/tcp", "2376"}, map[string][]string{"tcp": {"80"}, "udp": {"2377"}}},
+		{
+			name: "no change",
+			rule: &raw.Firewall{
+				Allowed: []*raw.FirewallAllowed{
+					{
+						IPProtocol: "tcp",
+						Ports:      []string{"80", "443"},
+					},
+				},
+			},
+			incomingPorts: []string{"443", "80"},
+			diffExpected:  false,
+		},
+		{
+			name: "add ports",
+			rule: &raw.Firewall{
+				Allowed: []*raw.FirewallAllowed{
+					{
+						IPProtocol: "tcp",
+						Ports:      []string{"80"},
+					},
+				},
+			},
+			incomingPorts: []string{"80/tcp", "443/tcp", "123/udp"},
+			diffExpected:  true,
+		},
+		{
+			name: "remove ports",
+			rule: &raw.Firewall{
+				Allowed: []*raw.FirewallAllowed{
+					{
+						IPProtocol: "tcp",
+						Ports:      []string{"80", "443"},
+					},
+					{
+						IPProtocol: "udp",
+						Ports:      []string{"123"},
+					},
+				},
+			},
+			incomingPorts: []string{"80/tcp"},
+			diffExpected:  true,
+		},
 	}
 
-	for _, test := range tests {
-		firewall := &raw.Firewall{Allowed: test.allowed}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			diff := updatePorts(tt.rule, tt.incomingPorts)
+			if diff && !tt.diffExpected {
+				t.Logf("expected change to be %t, but got %t", tt.diffExpected, diff)
+				t.Fail()
+			}
+
+			for _, p := range tt.incomingPorts {
+				port, proto := driverutil.SplitPortProto(p)
+				switch proto {
+				case "udp":
+					if !slices.Contains(tt.rule.Allowed[1].Ports, port) {
+						t.Logf("expected port %s to be in allowed list", port)
+						t.Fail()
+					}
+				default:
+					if !slices.Contains(tt.rule.Allowed[0].Ports, port) {
+						t.Logf("expected port %s to be in allowed list", port)
+						t.Fail()
+					}
+				}
+			}
+		})
+	}
+}
 
-		missingPorts := missingOpenedPorts(firewall, test.ports)
+func TestParseLabels(t *testing.T) {
+	var tests = []struct {
+		name     string
+		labels   string
+		expected map[string]string
+	}{
+		{
+			"empty",
+			"",
+			map[string]string{},
+		},
+		{
+			"valid label",
+			"one,two",
+			map[string]string{
+				"one": "two",
+			},
+		},
+		{
+			"valid labels",
+			"one,two,three,four",
+			map[string]string{
+				"one":   "two",
+				"three": "four",
+			},
+		},
+		{
+			"invalid format",
+			"one,two,three",
+			map[string]string{
+				"one": "two",
+			},
+		},
+	}
 
-		assert.Equal(t, test.expectedMissing, missingPorts, test.description)
+	t.Parallel()
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```

Comment:
```suggestion
	for _, tt := range tests {
	  tt := tt
		t.Run(tt.name, func(t *testing.T) {
```
IMO, move `t.Parallel()` to the top of the function if you are going to make it parallel. Also you need to shadow `tt`, otherwise you will have a race condition where tt can change in between tests so some cases may never be tested.